### DISCUSSION TITLE
Extending lines does not reduce subgraph/connected component counts w/ primal approach?

### DISCUSSION THREAD

**Discussion by tomrod** [pysal/momepy#356] [Q&A] [answered]
I have a series of linestrings that I read in using a GeoPandas GeoDataFrame. The geography linestrings have minor disconnects at several places. I am testing the `momepy.extend_lines` function before using the momepy.gdf_to_nx function. Unfortunately, no matter what I set tolerances and extents to, I still get way too many subgraphs / connected components on something that really should be a single graph.

Example code:

```python
gdf = gpd.read_file('filename',driver='GeoJSON').explode(index_parts=False)
tol = 1.e-3*min([
            abs(gdf.geometry.total_bounds[0] - gdf.geometry.total_bounds[2]),
            abs(gdf.geometry.total_bounds[1] - gdf.geometry.total_bounds[3])])
gdf2 = momepy.extend_lines(gdf, tolerance=tol, extension = 1/10 * tol)
G = momepy.gdf_to_nx(
        gdf2.to_crs('EPSG:4087'),  #originally in EPSG:4326
        approach='primal'
    )

nodes, edges = momepy.nx_to_gdf(G)
```

Checking `G` with `nx.number_connected_components(nx.to_undirected(G))` shows a few hundred subcomponents.

Additional notes

1. I have tested different tolerance levels (orders of magnitude). I confirmed that the visual output is indeed extending lines successfully, but this doesn't help the disconnectedness 
2. I have tried different CRS/EPSG without success
3. I have tested two different datasets with the same outcome -- in fact, the one that looks better comes back with more disconnected components!

Is there something in this process I am missing that would make the graph connect together with the primal approach?

**Comment by songololo** [comment]
@tomrod is the data source OpenStreetMap? Depending on the query you'll often end up with disconnected components from things like random walking paths in parks etc.

Are you able to see on plots what the nature of the disconnected components might be?

**Reply by tomrod** [reply]
Thanks so much for responding @songololo!

The data is not from OSM; it is a proprietary dataset of municipal infrastructure.

When I export both and plot the two layers overlapping, with the extensions in red and original in blue, I can clearly see the lines are extending. Image here (screenshot from QGIS:

![image](https://user-images.githubusercontent.com/1719700/165088208-e40b733c-9c15-4d36-9020-f03f40f57b86.png)

However, these are disconnected subcomponents after using mompey.gdf_to_nx with the `primal` approach.

**Comment by martinfleis** [comment, marked as answer]
Hi,

This is an expected behaviour. When creating networkx graph using `gdf_to_nx`, momepy infers topology of the network from the endpoints of the line strings. Which means that to create an intersection, endpoints of all line strings need to coincide at that point. If you just extend the lines to visually intersect the others, it does not create nodes there and treats such an intersection as non-planar, as if it was a bridge for example. 

I don't really have a universal solution for this case you are dealing with now. Snapping may be one option.

**Reply by tomrod** [reply]
Ok, that makes sense! I thought this might be the case but wasn't sure.

Is there a way you can recommend to create nodes at intersecting lines? You mentioned snapping but I wasn't sure how that resolves (my understanding is that snapping would move an item to an include geo, like a point to a line).

**Reply by tomrod** [reply]
This seems to be a pretty common item -- plenty of GeoPandas SO discussions (example: https://gis.stackexchange.com/questions/252474/intersecting-lines-to-get-points-in-geopandas). Feels like this is mostly there and once we get the intersections completed we'll be there. Thank you!
